keps/sig-cluster-lifecycle/kubeadm/20191125-remove-clusterstatus-from-kubeadm-config.md

```diff
@@ -0,0 +1,138 @@
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This KEP is proposing a new mode for tracking the list of the API endpoints in a cluster, thus allowing to remove the  `ClusterStatus` entry in the `kubeadm-config` ConfigMap and solve the problems that arise when, for any reasons, such entry does not reflect anymore the real status of the cluster.
+
+## Motivation
+
+In order to manage HA cluster properly, kubeadm requires to have access to the list of API endpoints in a cluster.
+
+Currently this feature is implemented by adding an entry in the list of API endpoints that is stored in the `ClusterStatus` entry of the `kubeadm-config` ConfigMap.
+
+There are well known problem related to the management of this list e.g. when a control-plane node dies or is deleted without invoking `kubeadm reset`, the list gets stale and the user is required to manually cleanup the list in order to avoid any kubeadm operation that relies on such list might incur into errors.
+
+This KEP is going to propose a different mode for tracking the list of the API endpoints in a cluster, based on the inspection of the current Pods.
+
+This approach does not require the maintenance of a separated list, and implicitly always reflect the current status of the cluster.
+
+This allows to remove the `ClusterStatus` entry in the `kubeadm-config` ConfigMap and to clean-up all the related code in `kubeadm init`, `kubeadm join` and `kubeadm reset`
+
+### Goals
+
+- To introduce a new method for tracking the list of API endpoints in a cluster
+- To allow removal of the `ClusterStatus` entry in the `kubeadm-config` ConfigMap and clean-up of the related goal.
+
+### Non-Goals
+
+- To change any user facing behavior in `kubeadm`
+
+## Proposal
+
+### Implementation Details/Notes/Constraints [optional]
+
+As of today, the `ClusterStatus` entry in the `kubeadm-config` ConfigMap contains a map that stores the `LocalAPIEndpoint` for each control-plane node.
+
+The `LocalAPIEndpoint` primary usage is for the `advertise-address` flag in the `kube-apiserver` pod. Having this value in a flag is not ideal for the purpose of this proposal, so, we are going to echo the same value into a new annotation named `kubeadm.kubernetes.io/kube-apiserver.advertise-address`.
+
+Once the annotation will be in place, it will be possible to easily retrieve the local advertise address for each control plane node by querying the corresponding `kube-apiserver` pod.
+
+The `LocalAPIEndpoint` is also used in the stacked `etcd` pod manifest for composing the `peer-urls` and the `client-urls`; the latter is used by kubeadm when accessing etcd in an existing cluster, e.g. when doing `join --control-plane`.
+
+We are going to echo the `client-urls` value into a new annotation named `kubeadm.kubernetes.io/etcd.advertise-client-urls`. Once the annotation will be in place, it will be possible to easily retrieve the etcd client urls by querying the `etcd` pods.
+
+### Risks and Mitigations
```

Comment:
I have not checked, but if a static pod goes down is that always removed from the pod list?  Also LBs would still need to update periodically.

Reply:
if the static manifest is removed the mirror Pod should be removed too.
but if the manifest is in place t would be possible do `kubectl describe po` on it to get the annotation.